app/cdap/components/PipelineDiff/util/diff.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { diff } from 'json-diff';
+import { DeepPartial, DiffIndicator, IPipelineDiffMap } from '../types';
+import {
+  IPipelineConfig,
+  IStageMap,
+  IConnectionMap,
+  IPipelineStage,
+  IPipelineConnection,
+} from '../types';
+
+export function getStageDiffKey(stage: IPipelineStage) {
+  return `${stage.plugin.name}__${stage.name}`;
+}
+
+export function getStageNameFromStageDiffKey(stageKey: string) {
+  return stageKey
+    .split('__')
+    .slice(1)
+    .join('__');
+}
+
+export function getPluginNameFromStageDiffKey(stageKey: string) {
+  return stageKey.split('__')[0];
+}
+
+export function getConnectionDiffKey(
+  connection: IPipelineConnection,
+  getStageDiffKeyFromPluginName: (name: string) => string
```

Comment:
why does this function need to be passed in? why can't you just pass in stageNameToStage and go stageNameToStage[connection.from] ?

Reply:
or i mean `stageNameToStage[getStageDiffKey[connection.from]]`

Reply:
It would be `getStageDiffKey(stageNameToStage[connection.from])`, and passing in stageNameToStage instead meant for the user of the function to construct a map each time even if they don't have to. I am not sure where else this function would be used, but it made sense to allow for other ways of getting the stage diff key from the stage name.